sync2/atxs.go

```diff
@@ -0,0 +1,326 @@
+		clock = clockwork.NewRealClock()
+	}
+	return &ATXHandler{
+		f:                f,
+		logger:           logger,
+		clock:            clock,
+		batchSize:        batchSize,
+		maxAttempts:      maxAttempts,
+		maxBatchRetries:  maxBatchRetries,
+		failedBatchDelay: failedBatchDelay,
+	}
+}
+
+func (h *ATXHandler) Commit(
+	ctx context.Context,
+	peer p2p.Peer,
+	base rangesync.OrderedSet,
+	received rangesync.SeqResult,
+) error {
+	h.logger.Debug("begin atx commit")
+	defer h.logger.Debug("end atx commit")
+	var firstK rangesync.KeyBytes
+	numDownloaded := 0
+	state := make(map[types.ATXID]int)
+	for k := range received.Seq {
+		if firstK == nil {
+			firstK = k
+		} else if firstK.Compare(k) == 0 {
+			break
+		}
+		found, err := base.Has(k)
+		if err != nil {
+			return fmt.Errorf("check if ATX exists: %w", err)
+		}
+		if found {
+			continue
+		}
+		id := types.BytesToATXID(k)
+		h.f.RegisterPeerHash(peer, id.Hash32())
+		state[id] = 0
+	}
+	if err := received.Error(); err != nil {
+		return fmt.Errorf("get item: %w", err)
+	}
+	total := len(state)
+	items := make([]types.ATXID, 0, h.batchSize)
+	startTime := h.clock.Now()
+	batchAttemptsRemaining := h.maxBatchRetries
+	for len(state) > 0 {
+		items = items[:0]
+		for id, n := range state {
+			if n >= h.maxAttempts {
+				h.logger.Debug("failed to download ATX: max attempts reached",
+					zap.String("atx", id.ShortString()))
+				delete(state, id)
+				continue
+			}
+			items = append(items, id)
+			if len(items) == h.batchSize {
+				break
+			}
+		}
+		if len(items) == 0 {
+			break
+		}
+
+		someSucceeded := false
+		var mtx sync.Mutex
+		err := h.f.GetAtxs(ctx, items, system.WithATXCallback(func(id types.ATXID, err error) {
+			mtx.Lock()
+			defer mtx.Unlock()
+			switch {
+			case err == nil:
+				numDownloaded++
+				someSucceeded = true
+				delete(state, id)
+			case errors.Is(err, pubsub.ErrValidationReject):
+				// if the atx invalid there's no point downloading it again
+				state[id] = h.maxAttempts
+			default:
+				state[id]++
+			}
+		}))
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				return err
+			}
+			if !errors.Is(err, &fetch.BatchError{}) {
+				h.logger.Debug("failed to download ATXs", zap.Error(err))
+			}
+		}
+		if !someSucceeded {
+			if batchAttemptsRemaining == 0 {
+				return errors.New("failed to download ATXs: max batch retries reached")
+			}
+			batchAttemptsRemaining--
+			h.logger.Debug("failed to download any ATXs: will retry batch",
+				zap.Int("remaining", batchAttemptsRemaining),
+				zap.Duration("delay", h.failedBatchDelay))
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-h.clock.After(h.failedBatchDelay):
+			}
+		} else {
+			batchAttemptsRemaining = h.maxBatchRetries
+			elapsed := h.clock.Since(startTime)
+			h.logger.Debug("fetched atxs",
+				zap.Int("total", total),
+				zap.Int("downloaded", numDownloaded),
+				zap.Float64("rate per sec", float64(numDownloaded)/elapsed.Seconds()))
+		}
```

Comment:
NIT: `else` isn't required 🙂 

```suggestion
		if !someSucceeded {
			if batchAttemptsRemaining == 0 {
				return errors.New("failed to download ATXs: max batch retries reached")
			}
			batchAttemptsRemaining--
			h.logger.Debug("failed to download any ATXs: will retry batch",
				zap.Int("remaining", batchAttemptsRemaining),
				zap.Duration("delay", h.failedBatchDelay),
			)
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-h.clock.After(h.failedBatchDelay):
				continue
			}
		}

		batchAttemptsRemaining = h.maxBatchRetries
		elapsed := h.clock.Since(startTime)
		h.logger.Debug("fetched atxs",
			zap.Int("total", total),
			zap.Int("downloaded", numDownloaded),
			zap.Float64("rate per sec", float64(numDownloaded)/elapsed.Seconds()),
		)
```

Reply:
Fixed